### PR TITLE
SI-6677 Insert required cast in `new qual.foo.T`

### PR DESCRIPTION
**Pull request by retronym**
Short version: we sometimes need to rewrite this as:

```
new T(qual.asInstanceOf[OwnerOfFoo].foo)
```

Long version:

`adaptMember` in Erasure performs a few tasks, among them:
1. adding an empty argument list qualifiers in `new qual.T`
   for which `qual` is a val template member that has
   (post uncurry) a MethodType with an empty parameter list.
   
   The same rewriting was already applied in uncurry for such
   qualifiers appearing in other context, e.g. `qual.foo` was
   already rewritten to `qual().foo`.
2. casting, if necessary, the qualifier in `Select(qual, name)`
   to the type of owner of the symbol that this selection
   references.
   
   This can be neccesary with compound types:
   - some val class member has type `A with B`;
   - we instantiate `new ab.valMemberOfB.T`
   - we must pass `ab.valMemberOfB` to the constructor of `T`
     as the `$outer` pointer
   - we must cast `ab` to `B` before calling `valMemberOfB`.
   
   Failure to insert this cast can lead to a LinkageError or
   a VerifyError.

However, if 1) was performed, 2) was not.

The error is pretty easy to trigger with the new reflection
API:

```
class Test {
  val cm: reflect.runtime.universe.Mirror
      = reflect.runtime.currentMirror
  def error {
    new cm.universe.Traverser // java.lang.VerifyError
  }

  def okay1 {
    val cm: reflect.runtime.universe.Mirror = reflect.runtime.currentMirror
    new cm.universe.Traverser
  }
}
```

The fix applied here to `adaptMember` mirrors the existing
implementation of `adaptType`.

Review by @odersky

### CODE REVIEW

**Comment by retronym** [comment]
PLS REBUILD ALL (it was just a comment change, but we can give it another spin...)

File: src/compiler/scala/tools/nsc/transform/Erasure.scala [hunk 1]
```diff
@@ -870,6 +875,8 @@ abstract class Erasure extends AddInterfaces
           tree1
       }
     }
+
+    private def isMethodTypeWithEmptyParams(tpe: Type) = tpe.isInstanceOf[MethodType] && tpe.params.isEmpty
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: I wonder if the temptation to write it the ugly way would be reduced if we could write

```
tpe matches { case MethodType(Nil, _) => }
```

Or if only patterns were first class,

```
tpe matches MethodType(Nil, _)
```

How do we get to that last one...

Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: https://groups.google.com/d/topic/scala-sips/zWB86iGVQcA/discussion

For now, I'll fix this.

**Comment by retronym** [comment]
@gkossakowski This one is ready to merge.

**Pull request merged**